source/particle/property/crystal_preferred_orientation.cc

```diff
@@ -0,0 +1,1160 @@
+                Assert(std::isfinite(vf_new),ExcMessage("vf_new is not finite before it is set. grain_i = "
+                                                        + std::to_string(grain_i) + ", volume_fractions[grain_i] = " + std::to_string(volume_fractions[grain_i])
+                                                        + ", derivatives.first[grain_i] = " + std::to_string(derivatives.first[grain_i])));
+
+                vf_new = volume_fractions[grain_i] + dt * vf_new * derivatives.first[grain_i];
+
+                Assert(std::isfinite(volume_fractions[grain_i]),ExcMessage("volume_fractions[grain_i] is not finite. grain_i = "
+                                                                           + std::to_string(grain_i) + ", volume_fractions[grain_i] = " + std::to_string(volume_fractions[grain_i])
+                                                                           + ", derivatives.first[grain_i] = " + std::to_string(derivatives.first[grain_i])));
+                if (std::fabs(vf_new-vf_old) < property_advection_tolerance)
+                  {
+                    break;
+                  }
+                vf_old = vf_new;
+              }
+
+            volume_fractions[grain_i] = vf_new;
+            sum_volume_fractions += volume_fractions[grain_i];
+          }
+
+        for (unsigned int grain_i = 0; grain_i < n_grains; ++grain_i)
+          {
+            auto cosine_old = rotation_matrices[grain_i];
+            auto cosine_new = rotation_matrices[grain_i];
+
+            for (size_t iteration = 0; iteration < property_advection_max_iterations; iteration++)
+              {
+                cosine_new = rotation_matrices[grain_i] + dt * cosine_new * derivatives.second[grain_i];
+
+                if ((cosine_new-cosine_old).norm() < property_advection_tolerance)
+                  {
+                    break;
+                  }
+                cosine_old = cosine_new;
+              }
+
+            rotation_matrices[grain_i] = cosine_new;
+          }
+
+        Assert(sum_volume_fractions != 0, ExcMessage("Sum of volumes is equal to zero, which is not supporsed to happen."));
+        return sum_volume_fractions;
+      }
+
+
+
+
+      template <int dim>
+      double
+      CrystalPreferredOrientation<dim>::advect_Crank_Nicolson(const double dt,
+                                                              const std::pair<std::vector<double>, std::vector<Tensor<2,3> > > &derivatives,
+                                                              std::vector<double> &volume_fractions,
+                                                              std::vector<Tensor<2,3> > &rotation_matrices,
+                                                              std::vector<double> &previous_volume_fraction_derivatives,
+                                                              std::vector<Tensor<2,3> > &previous_rotation_matrices_derivatives) const
+      {
+        double sum_volume_fractions = 0;
```

Comment:
Comments?

Reply:
I haven't really settled on one yet, but I have removed Crank Nicolson, since it is significantly more expensive an adds a lot of complexity. I think that it is good to keep the infrastructure to allow for multiple options in this case, so that Crank Nicolson can be added back if the need arises.